src/ape/pytest/fixtures.py

```diff
@@ -54,32 +69,114 @@ def project(self) -> ProjectManager:
     def _isolation(self) -> Iterator[None]:
         """
         Isolation logic used to implement isolation fixtures for each pytest scope.
+        When tracing support is available, will also assist in capturing receipts.
         """
-        snapshot_id = None
-        try:
-            snapshot_id = self.chain_manager.snapshot()
-        except ProviderNotConnectedError:
-            logger.warning("Provider became disconnected mid-test.")
 
-        except NotImplementedError:
-            if not self._warned_for_unimplemented_snapshot:
-                logger.warning(
-                    "The connected provider does not support snapshotting. "
-                    "Tests will not be completely isolated."
-                )
-                self._warned_for_unimplemented_snapshot = True
+        snapshot_id = self._snapshot()
 
-        yield
+        if self._using_traces:
+            with self.receipt_capture:
+                yield
+        else:
+            yield
 
-        if snapshot_id is not None and snapshot_id in self.chain_manager._snapshots:
-            try:
-                self.chain_manager.restore(snapshot_id)
-            except ProviderNotConnectedError:
-                logger.warning("Provider became disconnected mid-test.")
+        if snapshot_id:
+            self._restore(snapshot_id)
 
     # isolation fixtures
     _session_isolation = pytest.fixture(_isolation, scope="session")
     _package_isolation = pytest.fixture(_isolation, scope="package")
     _module_isolation = pytest.fixture(_isolation, scope="module")
     _class_isolation = pytest.fixture(_isolation, scope="class")
     _function_isolation = pytest.fixture(_isolation, scope="function")
+
+    @allow_disconnected
+    def _snapshot(self) -> Optional[SnapshotID]:
+        try:
+            return self.chain_manager.snapshot()
+        except NotImplementedError:
+            if not self._warned_for_unimplemented_snapshot:
+                logger.warning(
+                    "The connected provider does not support snapshotting. "
+                    "Tests will not be completely isolated."
+                )
+                self._warned_for_unimplemented_snapshot = True
+
+        return None
+
+    @allow_disconnected
+    def _restore(self, snapshot_id: SnapshotID):
+        if snapshot_id not in self.chain_manager._snapshots:
+            return
+
+        self.chain_manager.restore(snapshot_id)
+
+
+class ReceiptCapture(ManagerAccessMixin):
```

Comment:
Idea: was thinking we can potentially expose this class as a fixture for times when users want to track gas and manager their own snapshots:

```python
def test_something(receipt_capture, chain, contract, networks, owner):
    before_changing_code_snapshot = chain.snapshot()
    networks.provider.set_code(contract, HexBytes("0x123"))
    
    # Use receipt_capture so transaction not lost in gas report
    with receipt_capture:
        contact.doThing(sender=owner)
    
    chain.restore(before_changing_code_snapshot)
```

Reply:
Was hoping that this would always be the case. I could see wanting to opt-out of the behavior, but prefer it not be opt-in (because that will be confusing to have to opt-in to the behavior, instead you will have a disconnect in what's being captured)

Reply:
The other approach may be to add hooks to the `chain.snapshot()` call or something. Otherwise, not sure how you would be able to capture receipts if you you are managing snapshots within a test, and those receipts will get lost.

Reply:
Either way - I didn't do any of this  yet! Was just an idea. I think it is pretty rare users manager their own snapshots this way? I have seen it though.